woocommerce/woocommerce
# CRUD: populates `items` and `itemsCount` when a new item is created

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR populates the store's `items` and `itemsCount` properties when a new item is created. Depending on the case, it can improve the UX.
Important: the `optimistic` mode isn't enabled by default. It doesn't change the current behavior of the CRUD store.






Depends on https://github.com/woocommerce/woocommerce/issues/47623
Closes https://github.com/woocommerce/woocommerce/issues/47627



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

The PR is full of tests, trying to cover the most critical cases. Please feel free to take a look [here](https://github.com/woocommerce/woocommerce/blob/9a021263adf601a4b210b190aeb6441c70f1e538/packages/js/data/src/crud/test/utils.ts) and [here](https://github.com/woocommerce/woocommerce/blob/9a021263adf601a4b210b190aeb6441c70f1e538/packages/js/data/src/crud/test/reducer.ts).

You can run tests locally:

```cli
pnpm --filter=./packages/js/data run test:js packages/js/data/src/crud/test/
```

<img width="853" alt="Screenshot 2024-05-20 at 16 52 55" src="https://github.com/woocommerce/woocommerce/assets/77539/dff10296-74c0-4a61-9df8-128daafffee8">

You can use the client dev tools and the `data` package to test the.

1. Ensure to have installed the Redux dev tool ([Chrome](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?pli=1) / [Firefox](https://addons.mozilla.org/en-US/firefox/addon/reduxdevtools/))
2. Enable the Product editor
3. Create a new product
4. Open the dev console and the Redux dev tool
5. Filter the store by `/attributes`. Select the last one `wc/admin/products/attributes`:

<img width="760" alt="Screenshot 2024-05-20 at 17 01 53" src="https://github.com/woocommerce/woocommerce/assets/77539/6cd883cb-b863-41ab-88d0-82dfead0afbf">

6. In the dev console, dispatch an action to create a new global attribute. The snippet below creates a new `color` attribute. 

```es6
wp.data.dispatch( 'wc/admin/products/attributes' ).createProductAttribute( {
    name: 'color',
    generate_slug: true,
},
{
        optimisticQueryUpdate: { order_by: 'name' },
} )
```

7. Confirm that the `items` and `itemsCount` populate after the attribute is created:

<img width="667" alt="Screenshot 2024-05-20 at 17 04 53" src="https://github.com/woocommerce/woocommerce/assets/77539/7e810ec3-7cdc-4a34-a3b5-775677fa2d75">

8. You'd like to go to the Variations section to check the data propagation in the

<img width="830" alt="image" src="https://github.com/woocommerce/woocommerce/assets/77539/5f5748a6-4e4c-49d0-88c7-074ad591ccf9">

9. Pick the attribute ID recently created for use in the following steps. In my testing, it's `886`
10. Select the `wc/admin/products/attributes/terms` store:

<img width="534" alt="image" src="https://github.com/woocommerce/woocommerce/assets/77539/51c70af0-07be-4cfe-8c11-e58d4cd72ccb">

11. Dispatch a new action to create now an attribute term. Pass the attribute ID of the step 6

```es6
wp.data.dispatch( 'wc/admin/products/attributes/terms' ).createProductAttributeTerm( {
    name: 'pink',
    attribute_id: 886,
}, {
    optimisticQueryUpdate: {
        search: '',
        attribute_id: 886,
    },
    optimisticUrlParameters: [ 886 ],
} )
```

12. Confirm the `items` and `itemsCount` get populated:

<img width="531" alt="image" src="https://github.com/woocommerce/woocommerce/assets/77539/7898fa3f-ad95-42d9-95f4-43b706130f09">

13. You can confirm the new attribute in the options selector:

<img width="830" alt="image" src="https://github.com/woocommerce/woocommerce/assets/77539/397c3fa4-caa6-4772-8a69-1e29ede3342a">

### 🍿 


https://github.com/woocommerce/woocommerce/assets/77539/3a0e91a6-8c60-4a62-82af-42352e394567





### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
